Creating a WKWebExtension object from a zipped resource URL can sometimes fail

elijahsawyers opened the pull request:
#### 6813c5273edcc62401788a171d96b07b875ba37a
<pre>
Creating a WKWebExtension object from a zipped resource URL can sometimes fail
<a href="https://bugs.webkit.org/show_bug.cgi?id=284331">https://bugs.webkit.org/show_bug.cgi?id=284331</a>
<a href="https://rdar.apple.com/141179462">rdar://141179462</a>

Reviewed by Timothy Hatcher and Brian Weinstein.

This patch fixes a bug where creating a WKWebExtension object from a ZIP file
will fail if the contents of the zip are placed into a parent directory within
the temporary directory that&apos;s created.

The fix is to return the subdirectory of the temporary directory, if needed.

Also added a new test to cover this case.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):

Canonical link: <a href="https://commits.webkit.org/287639@main">https://commits.webkit.org/287639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3f37dcc3b092092e9813975ab5c01743f55a6c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80336 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59342 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33817 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84857 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31318 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7643 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62796 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20601 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73145 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43099 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50184 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27300 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29777 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86291 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71080 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70320 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13252 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12432 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7523 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7362 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10882 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9168 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -175,6 +175,15 @@ static String processFileAndExtractZipArchive(const String& path)
                 return;
             }
 
+            NSArray<NSString *> *contentsOfTemporaryDirectory = [NSFileManager.defaultManager contentsOfDirectoryAtPath:temporaryDirectory error:nil];
```
```suggestion
            auto *contentsOfTemporaryDirectory = [NSFileManager.defaultManager contentsOfDirectoryAtPath:temporaryDirectory error:nil];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -175,6 +175,15 @@ static String processFileAndExtractZipArchive(const String& path)
+                NSURL *temporaryDirectoryURL = [NSURL URLWithString:temporaryDirectory];
+                NSString *temporaryDirectorySubdirectoryPath = [temporaryDirectoryURL URLByAppendingPathComponent:contentsOfTemporaryDirectory.firstObject].path;
+
+                if ([NSFileManager.defaultManager fileExistsAtPath:temporaryDirectorySubdirectoryPath isDirectory:&isDirectory] && isDirectory)
```
```suggestion
                auto *temporaryDirectorySubdirectoryPath = [temporaryDirectory stringByAppendingPathComponent: contentsOfTemporaryDirectory.firstObject];
                if ([NSFileManager.defaultManager fileExistsAtPath:temporaryDirectorySubdirectoryPath isDirectory:&isDirectory] && isDirectory)
```

xeenon replied:
You could also get away with `subdirectoryPath` for the var name, since it is nested.

xeenon reviewed (approved):
You should add a new Zip file test that has a top-level folder to exercise this path.

elijahsawyers committed:
Creating a WKWebExtension object from a zipped resource URL can sometimes fail
https://bugs.webkit.org/show_bug.cgi?id=284331
rdar://141179462

Reviewed by Timothy Hatcher and Brian Weinstein.

This patch fixes a bug where creating a WKWebExtension object from a ZIP file
will fail if the contents of the zip are placed into a parent directory within
the temporary directory that's created.

The fix is to return the subdirectory of the temporary directory, if needed.

Also added a new test to cover this case.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):

Canonical link: https://commits.webkit.org/287639@main
